### ISSUE TITLE
Sample format 3 not available for device "hw:0,0": Invalid argument

### ISSUE THREAD

**Issue by skateskate** [mikebrady/shairport-sync#1087]
Hello, 
I have a problem with the following version of shairport-sync:
`3.1.6-OpenSSL-Avahi-ALSA-soxr-metadata-sysconfdir:/etc`

In particular, when I try to reproduce an audio file I get this error:
`fatal error: audio_alsa: Sample format 3 not available for device "hw:0,0": Invalid argument`
which corresponds to the S16 format.

The problem now is that if I list the specs of my audio card (hw:0,0), what I get is:
```
root@OpenWrt:/etc# cat /proc/asound/card0/stream0
C-Media Electronics Inc. USB 2.0 HD Audio at usb-ehci-platform-1.1, high speed : USB Audio

Playback:
  Status: Stop
  Interface 3
    Altset 1
    Format: S16_LE
    Channels: 2
    Endpoint: 4 OUT (ADAPTIVE)
    Rates: 44100
    Data packet interval: 1000 us
  Interface 3
    Altset 2
    Format: S24_3LE
    Channels: 2
    Endpoint: 4 OUT (ADAPTIVE)
    Rates: 44100
    Data packet interval: 1000 us
  Interface 3
    Altset 3
    Format: S16_LE
    Channels: 2
    Endpoint: 4 OUT (ADAPTIVE)
    Rates: 48000
    Data packet interval: 1000 us
  Interface 3
    Altset 4
    Format: S24_3LE
    Channels: 2
    Endpoint: 4 OUT (ADAPTIVE)
    Rates: 48000
    Data packet interval: 1000 us
  Interface 3
    Altset 5
    Format: S16_LE
    Channels: 2
    Endpoint: 4 OUT (ADAPTIVE)
    Rates: 96000
    Data packet interval: 1000 us
  Interface 3
    Altset 6
    Format: S24_3LE
    Channels: 2
    Endpoint: 4 OUT (ADAPTIVE)
    Rates: 96000
    Data packet interval: 1000 us
  Interface 3
    Altset 7
    Format: S16_LE
    Channels: 2
    Endpoint: 4 OUT (ADAPTIVE)
    Rates: 192000
    Data packet interval: 1000 us
  Interface 3
    Altset 8
    Format: S24_3LE
    Channels: 2
    Endpoint: 4 OUT (ADAPTIVE)
    Rates: 192000
    Data packet interval: 1000 us
```
so if I am not wrong, it should support S16 format. 
Do you know why I get the error? S16_LE format is not the same as S16? Sorry but this is my first time with linux/alsa...

**Comment by mikebrady** [comment]
Thanks for the interesting post. I see that you are using OpenWrt. Is the device itself big-endian or little-endian?

Older versions of Shairport Sync, including the version you are using, assumed, wrongly, that the endianness of the DAC would match the endianness of the CPU. So, if the machine was big-endian, it was assumed the DAC would accept S16BE.

Later versions of Shairport Sync have fixed that. Is 3.1.6 the latest available in OpenWrt?

**Comment by skateskate** [comment]
First of all thanks for the answer. I checked with the `lscpu` command and my device is big-endian. 
The latest version available in OpenWrt is the 3.2.2-3 at least to my knowledge. Unfortunately I have and old board which is poorly supported and in order to have another package (i.e., MPD, Music Player Daemon) to work, I had to downgrade all the build system to an older version... That is why I'm using the 3.1.6. 
At the moment I solved the S16 format problem using `plughw:0,0` as the output device but I don't know if it could have some drawbacks.

**Comment by mikebrady** [comment]
Thanks for the update. The `plughw` interface gives rate converters and adapters and is probably switching from S16BE to S16LE. Drawbacks are a little extra CPU demand and maybe a slight loss of exact speed. I found that the amount of drift was greater with it on a TP-710N.
